### PR TITLE
Add :for attr 

### PR DESCRIPTION
**Pull request by feliperenan**
Closes https://github.com/phoenixframework/phoenix_live_view/issues/2017

This PR adds the sugar `:for` as you can see below:

```eex
  <table id="my-table">
     <tr :for={user <- @users}>
       <td><%= user.name %>
     </tr>
   <table>
```

### TODO: 

- [x] handle self closed tags
- [x] raise when both for and let is present
- [x] raise when for is present on components (non normal HTML tags)
- [x] raise on duplicated :for
- [x] add docs
- [x] update Changelog

### CODE REVIEW

**Comment by josevalim** [comment]
> raise when both for and let is present

I think this is no longer necessary because you can use `:let` in tags and this PR raises for those cases.

File: lib/phoenix_live_view/html_engine.ex [hunk 1]
```diff
@@ -435,9 +435,21 @@ defmodule Phoenix.LiveView.HTMLEngine do
...
-    |> handle_tag_and_attrs(name, attrs, suffix, to_location(tag_meta))
+    case pop_special_attr!(attrs, "for", state) do
+      {{":for", expr, _meta}, attrs} ->
+        ast =
+          state
+          |> set_root_on_not_tag()
+          |> handle_tag_and_attrs(name, attrs, suffix, to_location(tag_meta))
+          |> handle_for_expr(expr)
+
+        update_subengine(state, :handle_expr, ["=", ast])
```
Comment (josevalim) [inline lib/phoenix_live_view/html_engine.ex:446]: You still need to call handle_begin in this case:

```suggestion
          state
          |> update_subengine(:handle_begin, [])
          |> set_root_on_not_tag()
          |> handle_tag_and_attrs(name, attrs, suffix, to_location(tag_meta))
          |> handle_for_expr(expr, state)
```

Then you can pass the old state to `handle_for_expr(...)`. `handle_for_expr()` can go ahead and call `        update_subengine(state, :handle_expr, ["=", ast])` on the old state too. :)

Comment (josevalim) [inline lib/phoenix_live_view/html_engine.ex:446]: You still need to call handle_begin in this case:

```suggestion
        state
        |> update_subengine(:handle_begin, [])
        |> set_root_on_not_tag()
        |> handle_tag_and_attrs(name, attrs, suffix, to_location(tag_meta))
        |> handle_for_expr(expr, state)
```

Then you can pass the old state to `handle_for_expr(...)`. `handle_for_expr()` can go ahead and call `        update_subengine(state, :handle_expr, ["=", ast])` on the old state too. :)

File: lib/phoenix_live_view/html_engine.ex [hunk 2]
```diff
@@ -446,19 +458,74 @@ defmodule Phoenix.LiveView.HTMLEngine do
...
   defp handle_token({:tag_close, name, tag_meta} = token, state) do
-    {{:tag_open, _name, _attrs, _tag_meta}, state} = pop_tag!(state, token)
-    update_subengine(state, :handle_text, [to_location(tag_meta), "</#{name}>"])
+    {{:tag_open, _name, _attrs, tag_open_meta}, state} = pop_tag!(state, token)
+    state = update_subengine(state, :handle_text, [to_location(tag_meta), "</#{name}>"])
+
+    case tag_open_meta[:for] do
+      {expr, outer_state} ->
+        ast = handle_for_expr(state, expr)
+        update_subengine(outer_state, :handle_expr, ["=", ast])
```
Comment (josevalim) [inline lib/phoenix_live_view/html_engine.ex:484]:
```suggestion
        handle_for_expr(state, expr, outer_state)
```

File: lib/phoenix_live_view/html_engine.ex [hunk 3]
```diff
@@ -505,6 +570,18 @@ defmodule Phoenix.LiveView.HTMLEngine do
     end)
   end
 
+  defp handle_for_expr(state, expr) do
+    expr = parse_expr!(expr, state)
+
+    quote do
+      for unquote(expr), do: unquote(invoke_subengine(state, :handle_end, []))
+    end
+  end
```
Comment (josevalim) [inline lib/phoenix_live_view/html_engine.ex:579]:
```suggestion
  defp handle_for_expr(state, expr, outer_state) do
    expr = parse_expr!(expr, state)

    ast =
      quote do
        for unquote(expr), do: unquote(invoke_subengine(state, :handle_end, []))
      end
    
    update_subengine(outer_state, :handle_expr, ["=", ast])
  end
```

File: lib/phoenix_live_view/html_engine.ex [hunk 4]
```diff
@@ -446,19 +456,73 @@ defmodule Phoenix.LiveView.HTMLEngine do
...
+    attrs
+    |> List.keytake(attr, 0)
+    |> raise_if_duplicated_special_attr!(state)
+  end
+
+  defp raise_if_duplicated_special_attr!({{attr, _expr, _meta}, attrs} = result, state) do
+    case List.keytake(attrs, attr, 0) do
+      {{attr, _expr, meta}, _attrs} ->
+        message =
+          "cannot define multiple #{inspect(attr)} attributes. Another #{inspect(attr)} has already been defined at line 3"
```
Comment (feliperenan) [inline lib/phoenix_live_view/html_engine.ex:510]:
```suggestion
          "cannot define multiple #{inspect(attr)} attributes. Another #{inspect(attr)} has already been defined at line #{meta.line}"
```

File: lib/phoenix_live_view/helpers.ex [hunk 5]
```diff
@@ -172,6 +172,40 @@ defmodule Phoenix.LiveView.Helpers do
...
+
+  ### HEEx extension: special attributes
+
+  Apart from normal HTML attributes, HEEx also support some special attributes
+  such as `:for` and `:let`.
+
+  #### :for
+
+  It is a syntax sugar for `<%= for .. do %>` that can used only in normal HTML
+  tags. Therefore `:for` will not work on components.
```
Comment (leandrocp) [inline lib/phoenix_live_view/helpers.ex:184]:
```suggestion
  It is a syntax sugar for `<%= for .. do %>` that can be used only in normal HTML
  tags, therefore `:for` will not work on components.
```

File: lib/phoenix_live_view/helpers.ex [hunk 6]
```diff
@@ -172,6 +172,40 @@ defmodule Phoenix.LiveView.Helpers do
...
+
+  ```eex
```
Comment (leandrocp) [inline lib/phoenix_live_view/helpers.ex:186]: To enable syntax highlighting

```suggestion
  ```heex
```

File: lib/phoenix_live_view/helpers.ex [hunk 7]
```diff
@@ -172,6 +172,40 @@ defmodule Phoenix.LiveView.Helpers do
...
+  <table>
+  ```
+
+  The snippet above will generate a list of `td` as you would expect.
+
+  #### :let
+
+  Useful for binding variables that will be used by the children elements.
+
+  ```eex
```
Comment (leandrocp) [inline lib/phoenix_live_view/helpers.ex:200]:
```suggestion
  ```heex
```

File: lib/phoenix_live_view/helpers.ex [hunk 8]
```diff
@@ -172,6 +172,40 @@ defmodule Phoenix.LiveView.Helpers do
...
+  <.form :let={f} for={@changeset} phx-change="validate" phx-submit="save">
+    <%= label f, :username %>
+    <%= text_input f, :username %>
```
Comment (leandrocp) [inline lib/phoenix_live_view/helpers.ex:203]:
```suggestion
    <%= label(f, :username) %>
    <%= text_input(f, :username) %>
```

That's how it gets formatted, I think it's worth using the same pattern on docs.

**Review by leandrocp — COMMENTED** [review]
A couple of suggestions on docs.

**Pull request merged**